## Input dữ liệu

In [400]:
import pandas as pd

In [401]:
# Import dữ liệu
link = "https://docs.google.com/spreadsheets/d/1ZNKbIUmS5apHKtI_WUXlwHeBkyu19wrt/edit?gid=1149835378"

In [402]:
# tách chuỗi link và chỉ lấy phần 1ZNKbIUmS5apHKtI_WUXlwHeBkyu19wrt
sheet_id = link.split("/d/")[1].split("/")[0]

sheets = ["customer", "ticket", "film"]
data = {}

for name in sheets:
    url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={name}"
    data[name] = pd.read_csv(url)

# Gọi từng sheet nếu cần
df_customer = data["customer"]
df_ticket = data["ticket"]
df_film = data["film"]

print(df_customer.head())


   Customerid         DOB Gender                 Address  Phường/Đường  \
0  0000000014   4/18/1999     Nữ  hoa khe quan thanh khe       Hòa Khê   
1  0000000034   4/16/1998    Nam                man thai      Mân Thái   
2  0000000051  12/16/1993    Nam       que son quang nam           NaN   
3  0000000081   11/8/1999    Nam            truong chinh  Trường Chinh   
4  0000000098   10/1/1994    Nam                  cẩm lệ           NaN   

  Quận/Huyện       Tỉnh       Website          job        industry  
0  Thanh Khê    Đà Nẵng   KH0104|0345      student        computer  
1    Sơn Trà    Đà Nẵng   KH0104|0333      student  health service  
2    Quế Sơn  Quảng Nam   KH0104|0255  blue collar       economics  
3     Cẩm Lệ    Đà Nẵng   KH0104|0293      student       economics  
4     Cẩm Lệ    Đà Nẵng  KH0104|40580  blue collar  health service  


In [403]:
# Kiểm tra thông tin của 3 sheet
print(df_customer.info())
print(df_film.info())
print(df_ticket.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4479 entries, 0 to 4478
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Customerid    4479 non-null   object
 1   DOB           4479 non-null   object
 2   Gender        4479 non-null   object
 3   Address       4478 non-null   object
 4   Phường/Đường  1298 non-null   object
 5   Quận/Huyện    4212 non-null   object
 6   Tỉnh          4464 non-null   object
 7   Website       3943 non-null   object
 8   job           4479 non-null   object
 9   industry      3357 non-null   object
dtypes: object(10)
memory usage: 350.1+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 26 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   show_id       76 non-null     object 
 1   title         76 non-null     object 
 2   director      65 non-null     object 
 3   cast          69 n

## Chuẩn bị dữ liệu

### Chuyển đối kiểu dữ liệu

In [404]:
# Bảng customer
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4479 entries, 0 to 4478
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Customerid    4479 non-null   object
 1   DOB           4479 non-null   object
 2   Gender        4479 non-null   object
 3   Address       4478 non-null   object
 4   Phường/Đường  1298 non-null   object
 5   Quận/Huyện    4212 non-null   object
 6   Tỉnh          4464 non-null   object
 7   Website       3943 non-null   object
 8   job           4479 non-null   object
 9   industry      3357 non-null   object
dtypes: object(10)
memory usage: 350.1+ KB


Có thể thấy DOB đang bị sai kiểu dữ liệu, chúng ta cần chuyển sang dữ liệu date

In [405]:
df_customer['DOB'] = pd.to_datetime(df_customer['DOB'], errors='coerce')

In [406]:
df_ticket.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35474 entries, 0 to 35473
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   orderid       35378 non-null  object 
 1   cashier       35474 non-null  object 
 2   saledate      35378 non-null  object 
 3   total         35378 non-null  float64
 4   customerid    35474 non-null  object 
 5   ticketcode    35474 non-null  int64  
 6   date          35474 non-null  object 
 7   time          35474 non-null  object 
 8   slot          35474 non-null  object 
 9   room          35474 non-null  int64  
 10  show_id       35474 non-null  object 
 11  slot type     35474 non-null  object 
 12  ticket type   35474 non-null  object 
 13  ticket price  35474 non-null  int64  
 14  popcorn       35378 non-null  object 
dtypes: float64(1), int64(3), object(11)
memory usage: 4.1+ MB


In [407]:
df_ticket.head(1)

orderid cashier  saledate    total  customerid  ticketcode  \
0  10006052019B0225  emp002  5/6/2019  90000.0  0000000014    20074925   

       date   time slot  room show_id slot type ticket type  ticket price  \
0  5/6/2019  18:15  G04     3    s966       ĐƠN  Thành viên         45000   

  popcorn  
0   Không

saledate, total, D/O/B, age, date,time đang bị sai kiểu dữ liệu

In [408]:
df_ticket['saledate'] = pd.to_datetime(df_ticket['saledate'], errors='coerce')
df_ticket['date'] = pd.to_datetime(df_ticket['date'], errors='coerce')
df_ticket['time'] = pd.to_datetime(df_ticket['time'], format='%H:%M', errors='coerce').dt.time
df_ticket['total'] = pd.to_numeric(df_ticket['total'], errors='coerce').fillna(0).astype(int)

In [409]:
# check thấy data này có nhiều cột dư thừa và cần loại bỏ
df_film = df_film.drop(df_film.columns[10:25], axis=1)

### Xử lý missing value

#### Bảng Customer

In [410]:
df_customer.head()

Customerid        DOB Gender                 Address  Phường/Đường  \
0  0000000014 1999-04-18     Nữ  hoa khe quan thanh khe       Hòa Khê   
1  0000000034 1998-04-16    Nam                man thai      Mân Thái   
2  0000000051 1993-12-16    Nam       que son quang nam           NaN   
3  0000000081 1999-11-08    Nam            truong chinh  Trường Chinh   
4  0000000098 1994-10-01    Nam                  cẩm lệ           NaN   

  Quận/Huyện       Tỉnh       Website          job        industry  
0  Thanh Khê    Đà Nẵng   KH0104|0345      student        computer  
1    Sơn Trà    Đà Nẵng   KH0104|0333      student  health service  
2    Quế Sơn  Quảng Nam   KH0104|0255  blue collar       economics  
3     Cẩm Lệ    Đà Nẵng   KH0104|0293      student       economics  
4     Cẩm Lệ    Đà Nẵng  KH0104|40580  blue collar  health service

In [411]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4479 entries, 0 to 4478
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Customerid    4479 non-null   object        
 1   DOB           4476 non-null   datetime64[ns]
 2   Gender        4479 non-null   object        
 3   Address       4478 non-null   object        
 4   Phường/Đường  1298 non-null   object        
 5   Quận/Huyện    4212 non-null   object        
 6   Tỉnh          4464 non-null   object        
 7   Website       3943 non-null   object        
 8   job           4479 non-null   object        
 9   industry      3357 non-null   object        
dtypes: datetime64[ns](1), object(9)
memory usage: 350.1+ KB


In [412]:
# Kiểm tra xem có bao nhiêu giá trị null
df_customer.isnull().sum()

Customerid         0
DOB                3
Gender             0
Address            1
Phường/Đường    3181
Quận/Huyện       267
Tỉnh              15
Website          536
job                0
industry        1122
dtype: int64

In [413]:
df_customer_industry_null = df_customer[df_customer['industry'].isna()]

In [414]:
df_customer_industry_null['job'].unique()

array(['teenager'], dtype=object)

Giá trị industry bị na ứng với nghề nghiệp là teenager, có thể fill giá trị na trong cột industry là ducation.

In [415]:
df_customer[['industry']] = df_customer[['industry']].fillna('education')

In [416]:
# Nhận thấy dữ liệu ở cột Address và Tỉnh bị miss số lượng nhỏ, không ảnh hưởng quá nhiều nếu phân phối dữ liệu, có thể xóa những giá trị đó khỏi tập dữ liệu.
df_customer = df_customer.dropna(subset = ["Tỉnh","Address","DOB"])
# Còn những dữ liệu bị thiếu ở cột khác, chuyển dữ liệu thành chưa biết, tránh ảnh hưởng đến phân phối dữ liệu
df_customer[["Phường/Đường","Quận/Huyện","Website"]] = df_customer[["Phường/Đường","Quận/Huyện","Website"]].fillna("Chưa biết")

C:\Users\Admin\AppData\Local\Temp\ipykernel_17492\739880111.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_customer[["Phường/Đường","Quận/Huyện","Website"]] = df_customer[["Phường/Đường","Quận/Huyện","Website"]].fillna("Chưa biết")


#### Bảng Ticket

In [417]:
df_ticket.isna().sum()

orderid         96
cashier          0
saledate        96
total            0
customerid       0
ticketcode       0
date             0
time             0
slot             0
room             0
show_id          0
slot type        0
ticket type      0
ticket price     0
popcorn         96
dtype: int64

In [418]:
# Nhận thấy giá trị bị null không ảnh hưởng nhiều đến phân phối dữ liệu, có thể loại bỏ, vì số lượng giá trị bị null chiếm phần nhỏ.
df_ticket = df_ticket.dropna(subset=["orderid"])

In [419]:
df_ticket.isna().sum()

orderid         0
cashier         0
saledate        0
total           0
customerid      0
ticketcode      0
date            0
time            0
slot            0
room            0
show_id         0
slot type       0
ticket type     0
ticket price    0
popcorn         0
dtype: int64

In [420]:
# Các giá trị bị null cùng một hàng liên kết với nhau, như trên, chỉ cần xóa 1 cột các giá trị null khác cũng được loại bỏ

#### Bảng Film

In [421]:
df_film.isna().sum()

show_id          0
title            0
director        11
cast             7
country          8
release_year     0
rating           9
duration         1
listed_in        1
description      1
Unnamed: 25     76
dtype: int64

In [422]:
# Do số lượng những giá trị null ít nhưng số hàng ở bảng film cũng khá ít, chúng em đã dựa trên thông tin của bộ phim để cố gắng fill vào chỗ na
df_film_director_null = df_film[df_film['director'].isnull()]

In [423]:
update_dict = {
    'Man vs. Shark': 'David Diley',
    'Lost City of Machu Picchu': 'Chưa biết',
    'Oil Spill of The Century': 'Fabrice Gardel, Josselin Mahot',
    "Star Wars: Galaxy's Edge-Adventure Awaits": 'Christian Lamb',
    'Women Of Impact: Changing The World': 'Lisa Feit',
    'Marvel Rising: Battle of The Bands': 'Chris Rutkowski',
    'Marvel Rising: Chasing Ghosts': 'Alfred Gimeno',
    'Marvel Rising: Heart of Iron': 'Sol Choi',
    'Happy Birthday, Mickey!': 'Bert Gillett',
    'Marvel Studios: Expanding the Universe': 'Jon Watts',
    '(PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU KÝ (G)': 'Yakuwa Shinnosuke' 
}

In [424]:
for title, director in update_dict.items():
    df_film.loc[
        (df_film['title'] == title) & (df_film['director'].isnull()),
        'director'
    ] = director


In [425]:
# Và khi chọn 1 bộ phim, bộ phim đến từ nước nào cũng rất quan trọng, đó là lý do nên fill cột country
df_film_country_null = df_film[df_film['country'].isnull()]
df_film_country_null

show_id                                              title  \
2     s194                     Disney My Music Story: Perfume   
4     s241                              Great Shark Chow Down   
7     s289                              Ultimate Viking Sword   
9     s313                           Oil Spill of The Century   
14    s433                                Chasing the Equinox   
42    s632                            Happy Birthday, Mickey!   
53    s963             Marvel Studios: Expanding the Universe   
73   s1244  (PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU K...   

                          director                         cast country  \
2               Kentaro Takayanagi  a-chan , KASHIYUKA , NOCCHi     NaN   
4                    Mark Woodward                 Dan Nachtrab     NaN   
7                       Peter Yost                  Jay Sanders     NaN   
9   Fabrice Gardel, Josselin Mahot                Loïck Peyron     NaN   
14                     Kenny Scott                 Stefan Frank     NaN   
42                    Bert Gillett                          NaN     NaN   
53                       Jon Watts                          NaN     NaN   
73               Yakuwa Shinnosuke                          NaN     NaN   

    release_year rating duration                         listed_in  \
2           2019    NaN   53 min  Biographical, Documentary, Music   
4           2019  TV-PG   45 min     Animals & Nature, Documentary   
7           2019  TV-PG   45 min           Documentary, Historical   
9           2019  TV-PG   45 min           Documentary, Historical   
14          2019  TV-PG   45 min                       Documentary   
42          2019   TV-G    2 min          Animation, Comedy, Dance   
53          2019  TV-PG   13 min               Documentary, Family   
73          2019    NaN      NaN                               NaN   

                                          description  Unnamed: 25  
2   J-Pop band Perfume shares their passion for mu...          NaN  
4   Great Shark Chow Down – prepare for a feast ...          NaN  
7   National Geographic reconstructs the Ulfberht,...          NaN  
9   Loïck Peyron investigates the 1978 Amoco Cadi...          NaN  
14  Ancient civilizations shared an incredible kno...          NaN  
42  Celebrating Mickey, the world’s most beloved...          NaN  
53  An exciting look into the future of Marvel Stu...          NaN  
73                                                NaN          NaN

In [426]:
update_country = {
    'Disney My Music Story: Perfume': 'Japan',
    'Great Shark Chow Down': 'United States',
    'Ultimate Viking Sword': 'United States',
    "Oil Spill of The Century": 'United States',
    'Chasing the Equinox': 'United States',
    'Happy Birthday, Mickey!': 'United States',
    'Marvel Studios: Expanding the Universe': 'United States',
    '(PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU KÝ (G)': 'Japan' 
}

In [427]:
for title, country in update_country.items():
    df_film.loc[
        (df_film['title'] == title) & (df_film['country'].isnull()),
        'country'
    ] = country

In [428]:
df_film_rating_null = df_film[df_film['rating'].isna()]

In [429]:
df_film_rating_null

show_id                                              title  \
2     s194                     Disney My Music Story: Perfume   
61   s1232                        NGÔI ĐỀN KỲ QUÁI  (Pee Nak)   
62   s1233                                  VÔ GIAN ĐẠO (C18)   
65   s1236               THẰNG EM LÝ TƯỞNG (Inseparable Bros)   
69   s1240  (LT) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU K...   
70   s1241                              LẬT MẶT: NHÀ CÓ KHÁCH   
72   s1243                       NỤ HÔN MA QUÁI (Sang Krasue)   
73   s1244  (PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU K...   
75   s1246                                CÀ CHỚN ANH ĐỪNG ĐI   

                             director  \
2                  Kentaro Takayanagi   
61       Phontharis Chotkijsadarsopon   
62                      Trần Việt Anh   
65                      Yook Sang-hyo   
69  Shinnosuke Yakuwa, Lee Jeong-beom   
70                             Lý Hải   
72               Sitisiri Mongkolsiri   
73                  Yakuwa Shinnosuke   
75                           Đỗ Cường   

                                                 cast   country  release_year  \
2                         a-chan , KASHIYUKA , NOCCHi     Japan          2019   
61  Chinawut Indracusin, Chutima Naina, Wachara Pa...  Thailand          2019   
62                     La Thành, Xuân Nghị, Huỳnh Anh   Vietnam          2019   
65                    Shin Ha Kyun,Lee Kwang Soo,Esom     Korea          2019   
69  Wasabi Mizuta, Megumi Oohara, Yumi Kakazu, Tom...     Japan          2019   
70                 Ly Hai, Huy Khanh, Thong, Jay Quan   Vietnam          2019   
72  Phantira Pipityakorn, Oabnithi Wiwattanawarang...  Thailand          2019   
73                                                NaN     Japan          2019   
75           Kiều Trinh, Xuân Phúc, Đức Hải, Trang Hí   Vietnam          2019   

   rating duration                         listed_in  \
2     NaN   53 min  Biographical, Documentary, Music   
61    NaN  108 min                    Comedy, Horror   
62    NaN  113 min           Comedy, Crime, Thriller   
65    NaN  114 min                     Comedy, Drama   
69    NaN  111 min      Animation, Action, Adventure   
70    NaN  100 min                    Comedy, Horror   
72    NaN  122 min            Drama, Horror, Romance   
73    NaN      NaN                               NaN   
75    NaN   90 min                           Romance   

                                          description  Unnamed: 25  
2   J-Pop band Perfume shares their passion for mu...          NaN  
61  Three men on their way to monk-hood, don't rea...          NaN  
62  Chay is a professional gambler. He and Dao, Th...          NaN  
65  Se-Ha is intelligent, but confined to a wheelc...          NaN  
69  Nobita saw a moon turned yellow last night and...          NaN  
70  When a woman and her friends visit her home vi...          NaN  
72  In a far-away village, lived an innocent teena...          NaN  
73                                                NaN          NaN  
75  Tue Nhi, who studied music composition, is hav...          NaN

In [430]:
update_rating = {
    'Disney My Music Story: Perfume': 'TV-G',
    'NGÔI ĐỀN KỲ QUÁI  (Pee Nak)': 'PG-13',
    'VÔ GIAN ĐẠO (C18)': 'C18',
    "THẰNG EM LÝ TƯỞNG (Inseparable Bros)": 'PG-13',
    "(LT) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU KÝ (Doraemon: Nobita's Chronicle of the Moon Exploration)": 'G',
    'LẬT MẶT: NHÀ CÓ KHÁCH': 'C16',
    'NỤ HÔN MA QUÁI (Sang Krasue)': 'PG-13',
    '(PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU KÝ (G)': 'G',
    'CÀ CHỚN ANH ĐỪNG ĐI': 'C13-C16'
}

In [431]:
for title, rating in update_rating.items():
    df_film.loc[
        (df_film['title'] == title) & (df_film['rating'].isnull()),
        'rating'
    ] = rating

In [432]:

df_film_listed_in_null = df_film[df_film['listed_in'].isna()]

In [433]:
df_film_listed_in_null

show_id                                              title  \
73   s1244  (PĐ) DORAEMON: NOBITA VÀ MẶT TRĂNG PHIÊU LƯU K...   

             director cast country  release_year rating duration listed_in  \
73  Yakuwa Shinnosuke  NaN   Japan          2019      G      NaN       NaN   

   description  Unnamed: 25  
73         NaN          NaN

In [434]:
# Bộ phim doraemon có thể loại: Animation, Action, Adventure
df_film[['listed_in']] = df_film[['listed_in']].fillna('Animation, Action, Adventure')

Còn một số cột bị thiếu nhưng chúng em nhận thấy nó không ảnh hưởng đến hành vi mua hàng của khách hàng, nên chúng tôi vẫn giữ nguyên và chỉ dùng những cột quan trọng của bộ data này.

Trước khi tiến hành feature engineering, để thuận tiện hơn cho việc visualize cũng như các bước phân tích sau đó, chúng em sẽ quyết định gộp 3 file lại thành 1 file final.

Trong đó:
- Bảng customer: Customerid là khóa chính merged với cột customerid bên bảng ticket
- Bảng film: show_id là khóa chính và merged với cột show_id bên bảng ticket


In [435]:
# Merged ticket với customer
merged_ticket_customer = pd.merge(df_ticket,df_customer, left_on = 'customerid', right_on = 'Customerid',how = 'left')

In [436]:
df_final = pd.merge(merged_ticket_customer,df_film, on = 'show_id',how = 'left' )

In [437]:
df_final.isna().sum()

orderid             0
cashier             0
saledate            0
total               0
customerid          0
ticketcode          0
date                0
time                0
slot                0
room                0
show_id             0
slot type           0
ticket type         0
ticket price        0
popcorn             0
Customerid         95
DOB                95
Gender             95
Address            95
Phường/Đường       95
Quận/Huyện         95
Tỉnh               95
Website            95
job                95
industry           95
title               0
director            0
cast               49
country             0
release_year        0
rating              0
duration           49
listed_in           0
description        49
Unnamed: 25     35378
dtype: int64

Những cột Customerid, DOB, Gender,... như quan sát trên có số hàng miss value là như nhau, có thể do có thông tin vé mua nhưng mà không có thông tin khách hàng, số lượng missing value cũng không lớn, có tầm 95 hàng, rất ít so với 35378 quan sát, nên có thể loại bỏ các hàng này.

Cùng với đó chúng em tiến hành loại bỏ những cột mà không cần dùng đến để tranh gây rối, khó quan sát.


In [438]:
columns_to_check = ['Customerid', 'DOB', 'Gender', 'Address', 'Phường/Đường',
                    'Quận/Huyện', 'Tỉnh', 'Website', 'job', 'industry']

df_final = df_final.dropna(subset=columns_to_check)


In [439]:
columns_to_drop = ['cast', 'release_year', 'duration', 'description', 'Unnamed: 25', 'Customerid','Website','Phường/Đường','Address']
df_final.drop(columns=columns_to_drop, inplace=True)

In [440]:
df_final.isna().sum()

orderid         0
cashier         0
saledate        0
total           0
customerid      0
ticketcode      0
date            0
time            0
slot            0
room            0
show_id         0
slot type       0
ticket type     0
ticket price    0
popcorn         0
DOB             0
Gender          0
Quận/Huyện      0
Tỉnh            0
job             0
industry        0
title           0
director        0
country         0
rating          0
listed_in       0
dtype: int64

In [441]:
df_final['Tỉnh'].value_counts()

Tỉnh
Đà Nẵng          34032
Quảng Nam          796
Huế                129
Quảng Bình          71
Quảng Trị           58
Quảng Ngãi          47
Gia Lai             20
Hà Nội              18
Hà Tĩnh             16
Đắk Lắk             14
Thanh Hóa           13
Ninh Bình           11
Kon Tum             11
Bình Định           10
Hồ Chí Minh         10
Nghệ An              7
Kiên Giang           4
Bắc Giang            3
Buôn Mê Thuộc        2
Lạng Sơn             2
Phú Yên              2
Đồng Nai             2
Thái Bình            2
Dak Lak              2
Yên Bái              1
Name: count, dtype: int64

Dữ liệu chủ yếu là từ Đà Nẵng, để có cái nhìn toàn diện hơn, xóa các tỉnh thành khác, chỉ giữ lại Đà Nẵng để xem doanh thu, xu hướng mua vé của khách hàng tại Đà Nẵng.

In [442]:
df_final = df_final[df_final['Tỉnh'] == 'Đà Nẵng']

In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 34032 entries, 0 to 35377
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   orderid               34032 non-null  object        
 1   cashier               34032 non-null  object        
 2   saledate              34032 non-null  datetime64[ns]
 3   total                 34032 non-null  int64         
 4   customerid            34032 non-null  object        
 5   ticketcode            34032 non-null  int64         
 6   date                  34032 non-null  datetime64[ns]
 7   time                  34032 non-null  datetime64[ns]
 8   slot                  34032 non-null  object        
 9   room                  34032 non-null  int64         
 10  show_id               34032 non-null  object        
 11  slot type             34032 non-null  object        
 12  ticket type           34032 non-null  object        
 13  ticket price         

Sau khi đã merge các bảng lại với nhau và cũng như đã xử ly dữ liệu, dữ liệu đã sẵn sàng được đưa vào phân tích, cũng như visualization.

## Feature Engineering

In [443]:
df_final.head(5)

orderid cashier   saledate  total  customerid  ticketcode  \
0  10006052019B0225  emp002 2019-05-06  90000  0000000014    20074925   
1  10006052019B0225  emp002 2019-05-06  90000  0000000014    20074924   
2  10022052019B0167  emp011 2019-05-22  90000  0000000034    20097042   
3  10022052019B0167  emp011 2019-05-22  90000  0000000034    20097041   
4  10130052019B0141  emp005 2019-05-30  90000  0000000034    20106761   

        date      time slot  room  ... Gender Quận/Huyện     Tỉnh      job  \
0 2019-05-06  18:15:00  G04     3  ...     Nữ  Thanh Khê  Đà Nẵng  student   
1 2019-05-06  18:15:00  G03     3  ...     Nữ  Thanh Khê  Đà Nẵng  student   
2 2019-05-22  20:45:00  E04     2  ...    Nam    Sơn Trà  Đà Nẵng  student   
3 2019-05-22  20:45:00  E03     2  ...    Nam    Sơn Trà  Đà Nẵng  student   
4 2019-05-30  19:30:00  F04     1  ...    Nam    Sơn Trà  Đà Nẵng  student   

         industry                                    title  \
0        computer        Marvel Studios' Avengers: Endgame   
1        computer        Marvel Studios' Avengers: Endgame   
2  health service  JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)   
3  health service  JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)   
4  health service              NGÔI ĐỀN KỲ QUÁI  (Pee Nak)   

                       director        country rating  \
0      Anthony Russo, Joe Russo  United States  PG-13   
1      Anthony Russo, Joe Russo  United States  PG-13   
2                Chad Stahelski  United States      R   
3                Chad Stahelski  United States      R   
4  Phontharis Chotkijsadarsopon       Thailand  PG-13   

                                    listed_in  
0  Action-Adventure, Fantasy, Science Fiction  
1  Action-Adventure, Fantasy, Science Fiction  
2                                      Action  
3                                      Action  
4                              Comedy, Horror  

[5 rows x 26 columns]

Sau khi quan sát dữ liệu, có thể thêm một số cột sau:
- Tuổi: tính tuổi khách hàng
- Khoảng tuối, chia tuổi khách hàng thành nhiều nhóm tuổi khác nhau
- Thời gian trong ngày, sáng, chiều

In [444]:
# Đảm bảo cả 'DOB' và 'saledate' đều là kiểu datetime
df_final['DOB'] = pd.to_datetime(df_final['DOB'], errors='coerce')
df_final['saledate'] = pd.to_datetime(df_final['saledate'], errors='coerce')

# Tính tuổi tại thời điểm mua hàng
df_final['Tuổi'] = (df_final['saledate'] - df_final['DOB']).dt.days // 365


In [ ]:
df_final['time'] = pd.to_datetime(df_final['time'], format='%H:%M:%S', errors='coerce')
df_final['Thời gian trong ngày'] = df_final['time'].dt.hour.apply(time_of_day)
def time_of_day(hour):
    if 5 <= hour < 12:
        return 'Sáng'
    elif 12 <= hour < 17:
        return 'Chiều'
    elif 17 <= hour < 21:
        return 'Tối'
    else:
        return 'Đêm'

df_final['Thời gian trong ngày'] = df_final['time'].dt.hour.apply(time_of_day)


In [458]:
df_final['Tuổi'].value_counts()

Tuổi
 0     10137
 18     8468
 19     1432
 20     1398
 22     1393
 21     1299
 23     1171
 24     1085
 17      965
 25      799
 16      791
 26      713
 27      605
 15      541
 1       534
 28      449
 29      314
 14      237
 30      223
 13      207
 31      189
 34      124
 33       97
 32       88
 37       83
 35       82
 36       73
 39       72
 40       45
 38       44
 11       38
 2        37
 12       35
 42       34
 41       33
 3        27
 50       24
 51       19
 48       18
 43       17
 44       17
 45       17
 10       16
 47       15
 49        9
 54        8
 52        3
 53        3
-1         2
-76        1
 55        1
Name: count, dtype: int64

In [459]:
df_final = df_final[~df_final['Tuổi'].isin([-1, -76])]

In [467]:
df_tre = df_final[df_final['Tuổi'].isin([0,1,2,3])]
df_tre['job'].unique()

array(['teenager'], dtype=object)

Đang có sự sai lệch tuổi nghiêm trọng, các job là teenager nhưng vẫn có độ tuổi 0,1,2,3, tuy nhiên số lượng chúng khá lớn, nên cách tốt nhất là để những quan sát như vậy làm một nhóm tuổi riêng.

In [461]:
def group_age(age):
    if age == 0:
        return '0 tuổi'
    elif age == 1:
        return '1 tuổi'
    elif age == 2:
        return '2 tuổi'
    elif age == 3:
        return '3 tuổi'
    elif 4 <= age <= 11:
        return 'Trẻ em (4-11)'
    elif 12 <= age <= 17:
        return 'Thiếu niên (12-17)'
    elif 18 <= age <= 25:
        return 'Thanh niên (18-25)'
    elif 26 <= age <= 40:
        return 'Trưởng thành (26-40)'
    elif 41 <= age <= 60:
        return 'Trung niên (41-60)'
    else:
        return 'Cao tuổi (trên 60)'
df_final['Nhóm tuổi'] = df_final['Tuổi'].apply(group_age)
